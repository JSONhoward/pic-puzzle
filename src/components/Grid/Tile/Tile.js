import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import Arrows from './Arrows/Arrows'
import { useResetRecoilState, useSetRecoilState } from 'recoil'
import { ShowArrows, puzzleImage, TileContent } from '../../../Store'
import { useRecoilValue } from 'recoil'

const StyledTile = styled.div`
position: relative;
display: grid;
place-items: center;
color: white;
grid-column: ${props => props.column};
grid-row: ${props => props.row};
border: 1px solid white;
overflow: hidden;
`

const ContentImage = styled.div`
position: relative;
height: 100%;
width: 100%;
background-image: ${props => props.tile !== null ? 'url(' + props.bg + ')' : 'none'};
background-size: 90vh;
background-position: ${props => {
        switch (props.tile) {
            case 1:
                return '0 0'
            case 2:
                return '50% 0'
            case 3:
                return '100% 0'
            case 4:
                return '0 50%'
            case 5:
                return '50% 50%'
            case 6:
                return '100% 50%'
            case 7:
                return '0 100%'
            case 8:
                return '50% 100%'
            default:
                return '0 0'
        }
    }};

@media screen and (orientation: portrait) {
    background-size: 90vw;
}
`

const Tile = ({ column, row, content, show, tileId }) => {
    const resetArrows = useResetRecoilState(ShowArrows)
    const tileImage = useRecoilValue(puzzleImage)
    const tileContent = useRecoilValue(TileContent)
    const setShowArrow = useSetRecoilState(ShowArrows)

    const handleMouseEnter = (e) => {
        const tileNum = e.target.getAttribute('name')
        resetArrows()
        setShowArrow(prevVal => {
            if (tileContent[`tile${tileNum}`] !== null) {
                let state = {
                    ...prevVal,
                    [`tile${tileNum}`]: true
                }
                return state
            }
            return prevVal
        })
    }

    return (
        <StyledTile tileId={tileId} column={column} row={row}>
            <ContentImage name={tileId} onMouseEnter={(e) => handleMouseEnter(e)} onMouseLeave={resetArrows} bg={tileImage} tile={content}>
                {show && <Arrows tileNum={tileId} column={column} row={row} />}
            </ContentImage>
        </StyledTile>
    )
}

Tile.propTypes = {
    column: PropTypes.number.isRequired,
    row: PropTypes.number.isRequired,
    content: PropTypes.node,
    show: PropTypes.bool.isRequired,
    tileId: PropTypes.number.isRequired
}

export default Tile
