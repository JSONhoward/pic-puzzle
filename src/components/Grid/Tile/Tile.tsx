import React from 'react'
import styled from 'styled-components'
import { useResetRecoilState, useSetRecoilState, useRecoilValue } from 'recoil'

import Arrows from './Arrows/Arrows'
import { ShowArrows, puzzleImage, TileContent, TileMoves } from '../../../Store'

const StyledTile = styled('div')<{col: any, row: any}>`
position: relative;
display: grid;
place-items: center;
color: white;
grid-column: ${props => props.col};
grid-row: ${props => props.row};
overflow: hidden;
`

const EmptyTile = styled.div`
display: grid;
place-items: center;
height: 100%;
width: 100%;
font-size: 2rem;
background-color: rgb(30,30,30);
`

const ContentImage = styled('div')<{tile: Number, bg: String}>`
position: relative;
height: 100%;
width: 100%;
background-image: ${props => props.tile !== null ? 'url(' + props.bg + ')' : 'none'};
background-size: 80vh;
background-position: ${props => {
        switch (props.tile) {
            case 1:
                return '0 0'
            case 2:
                return '50% 0'
            case 3:
                return '100% 0'
            case 4:
                return '0 50%'
            case 5:
                return '50% 50%'
            case 6:
                return '100% 50%'
            case 7:
                return '0 100%'
            case 8:
                return '50% 100%'
            default:
                return '0 0'
        }
    }};

@media screen and (orientation: portrait) {
    background-size: 80vw;
}
`

type TileProps = {
    column: Number,
    row: Number,
    content: Number | null,
    show: Boolean,
    tileId: Number
}

const Tile: React.FC<TileProps> = ({ column, row, content, show, tileId }) => {
    const resetArrows = useResetRecoilState(ShowArrows)
    const tileImage = useRecoilValue<String>(puzzleImage)
    const tileContent = useRecoilValue<Object>(TileContent)
    const setShowArrow = useSetRecoilState<Boolean | any>(ShowArrows)
    const moveCount = useRecoilValue(TileMoves)

    const handleMouseEnter = (e: React.MouseEvent<HTMLDivElement>) => {
        let element = e.target as HTMLElement
        const tileNum = element.getAttribute('name')
        const tileContentProperty = (tileContent as {[key: string]: any})[`tile${tileNum}`]
        resetArrows()
        setShowArrow((prevVal: any) => {
            if (tileContentProperty !== null) {
                let state = {
                    ...prevVal,
                    [`tile${tileNum}`]: true
                }
                return state
            }
            return prevVal
        })
    }

    return (

        <StyledTile col={column} row={row}>
            {
                content === null ?
                    (
                        <>
                            <EmptyTile>
                            <p>Moves: {moveCount}</p>
                            </EmptyTile>
                        </>
                    )
                    :
                    (
                        <ContentImage onMouseEnter={(e: React.MouseEvent<HTMLDivElement>): void => handleMouseEnter(e)} onMouseLeave={resetArrows} bg={tileImage} tile={content}>
                            {show && <Arrows tileNum={tileId} column={column} row={row} />}
                        </ContentImage>
                    )
            }
        </StyledTile>

    )
}

export default Tile
