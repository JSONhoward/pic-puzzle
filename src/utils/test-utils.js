import React from 'react'
import renderer from 'react-test-renderer'
import { RecoilRoot } from 'recoil'

export const renderWrapper = (elements) => {
    return renderer.create(<RecoilRoot>{elements}</RecoilRoot>)
}