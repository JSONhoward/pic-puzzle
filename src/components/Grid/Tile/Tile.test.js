import React from 'react'
import Tile from './Tile'
import {renderWrapper} from '../../../utils/test-utils'

describe('Tile', () => {
    test('snapshot renders', () => {
        const component = renderWrapper(<Tile />)
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
})