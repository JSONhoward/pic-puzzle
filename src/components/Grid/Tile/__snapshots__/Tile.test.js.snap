// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile snapshot renders 1`] = `
<div
  className="sc-AxiKw kDujED"
>
  <div
    className="sc-AxhUy iNZIgF"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  />
</div>
`;
